### PR TITLE
Verify wallet before delete

### PR DESCRIPTION
**Pull request by amanCypherock**
This PR incorporates the change where a wallet is verified before it is deleted.

These are the changes incorporated
- Record card write attempt on flash
- Fetch share (and compare with wallet intended for deletion) from the card before deleting the wallet
- Updated factory reset to include all wallets with share present on device during verification from 2 cards.

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-6557 Verify wallet nonce on fetching a wallet](https://app.clickup.com/t/9002019994/PRF-6557)

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-6558 Verify wallet before deleting](https://app.clickup.com/t/9002019994/PRF-6558)

File: src/card_operations/card_write_share.c [hunk 1]
```diff
@@ -128,16 +137,31 @@ static void write_card_pre_process(uint8_t card_num) {
   return;
 }
 
-static void write_card_share_post_process(uint8_t card_num) {
+static void record_card_write_attempt_on_flash(uint8_t card_num) {
   Flash_Wallet *wallet_for_flash = get_flash_wallet();
-  wallet_for_flash->cards_states = (1 << card_num) - 1;
+  uint8_t temp = encode_card_number(card_num) << 4;
+
+  if (temp == (temp & wallet_for_flash->cards_states)) {
```
Comment (ujjwal-cyph) [inline src/card_operations/card_write_share.c:144]: comment the reasoning here

File: common/interfaces/flash_interface/flash_api.c [hunk 2]
```diff
@@ -651,9 +653,10 @@ bool card_already_deleted_flash(const uint8_t index,
   ASSERT(index < MAX_WALLETS_ALLOWED);
 
   get_flash_ram_instance();    // to load
-  return !(
-      ((flash_ram_instance.wallets[index].cards_states) >> (card_number - 1)) &
-      1);
+  uint8_t card_state =
+      (flash_ram_instance.wallets[index].cards_states & 0x0F) |
+      (flash_ram_instance.wallets[index].cards_states >> 4 & 0x0F);
+  return !(0x01 == (0x01 & (card_state >> (card_number - 1))));
```
Comment (ujjwal-cyph) [inline common/interfaces/flash_interface/flash_api.c:659]: add explanation in comments

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]
fix minor comments

File: src/card_operations/card_write_share.c (hunk 1 above)
Comment (amanCypherock) [reply src/card_operations/card_write_share.c]: added

File: common/interfaces/flash_interface/flash_api.c (hunk 2 above)
Comment (amanCypherock) [reply common/interfaces/flash_interface/flash_api.c]: refactored

**Pull request merged**